File: Baekjoon/[ 17472 ] 다리 만들기2 [ 190907 SW역량테스트 ].py
# https://www.acmicpc.net/problem/17472

def dfs(r,c,idx):       # 이름 지어주기
    for dr, dc in dir:
        rr = r + dr
        cc = c + dc
        if 0 <= rr and rr < R and 0 <= cc and cc < C:    
            if arr[rr][cc] == 1:
                arr[rr][cc] = idx
                dfs(rr,cc,idx)

def bfs(idx):           # idx 섬으로 부터 각 섬까지의 거리
    que = []
    for r in range(R):
        for c in range(C):
            if arr[r][c] == idx:
                for d in range(4):
                    rr = r + dir[d][0]
                    cc = c + dir[d][1]
                    if 0 <= rr and rr < R and 0 <= cc and cc < C:
                        que.append([rr,cc, d])
    count = 0
    while que:
        que_ = []

        for r,c,d in que:
            rr = r + dir[d][0]
            cc = c + dir[d][1]
            if arr[r][c] == 0:
                if 0 <= rr and rr < R and 0 <= cc and cc < C:
                    que_.append([rr,cc,d])
            else:
                if arr[r][c] == idx:
                    continue
                if count < 2:
                    continue
                if inner[idx][arr[r][c]] == 0:
                    inner[idx][arr[r][c]] = count

        que = que_
        count += 1


dir = [[-1,0], [1,0], [0,-1], [0,1]]
R, C = list(map(int, input().split()))
arr = [list(map(int, input().split())) for _ in range(R)]

idx = 2
for r in range(R):
    for c in range(C):
        if arr[r][c] == 1:
            arr[r][c] = idx
            dfs(r,c, idx)
            idx += 1

inner = [[0] * idx for _ in range(idx)]     # 인접 배열

for i in range(2,idx):                      #  i 섬에서 다른 섬까지의 거리
    bfs(i)

brr = []                                    # A, B, V : A에서 B까지의 거리는 V

for r in range(2,idx):
    for c in range(r+1,idx):
        if inner[r][c] == 0:
            continue
        brr.append([r,c,inner[r][c]])

brr = sorted(brr, key= lambda x: x[2])

crr = []
visit = [-1] * idx
answer = 0

for a, b, v in brr:                         # 유니온같은거

    va = visit[a]
    vb = visit[b]
    if va == -1 and vb == -1:
        visit[a] = len(crr)
        visit[b] = len(crr)
        crr.append([a,b])
    elif va == -1:       # va 만 0
        visit[a] = vb
        crr[vb].append(a)
    elif vb == -1:       # vb 만 0
        visit[b] = va
        crr[va].append(b)
    else:
        if va == vb:
            continue
        else:
            if va < vb: # vb > va
                for b_ in crr[vb]:
                    visit[b_] = va
                crr[va] += crr[vb]
                crr[vb] = []
            else:   # va > vb
                for a_ in crr[va]:
                    visit[a_] = vb
                crr[vb] += crr[va]
                crr[va] = []
    answer += v

if crr:
    if len(crr[0]) != idx-2:
        print('-1')
    else:
        print(answer)
else:
    print('-1')

File: Baekjoon/[ 17472 ] 다리 만들기2 [ 190907 SW역량테스트 ].java
package 다리만들기2;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    /*
     1 <= R, C <= 10
     3 <= R * C <= 100
     2 <= 섬 <= 6
     */
    static class Loc{
        int r;
        int c;
        int d; // 방향
        Loc(int r, int c){
            this.r = r;
            this.c = c;
        }
        Loc(int r, int c, int d){
            this.r = r;
            this.c = c;
            this.d = d;
        }
    }

    static class Dist implements Comparable<Dist>{
        int s;
        int e;
        int d;
        Dist(int s, int e, int d){
            this.s = s;
            this.e = e;
            this.d = d;
        }

        @Override
        public int compareTo(Dist dist) {
            return this.d - dist.d;
        }

        @Override
        public String toString() {
            return "[ " + s + " , " + e + " , " + d + " ] ";
        }
    }

    static int R;
    static int C;
    static int[][] map;
    static int[][] adj;
    static boolean[][] visit;
    static int idx;
    static int[][] dir = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    static LinkedList<Dist> dists;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String[] line = sc.nextLine().split(" ");
        R = Integer.parseInt(line[0]);
        C = Integer.parseInt(line[1]);
        map = new int[R][C];
        for(int r = 0 ; r<R; r++){
            map[r] = new int[C];
            line = sc.nextLine().split(" ");
            for(int c = 0; c <C; c++){
                map[r][c] = Integer.parseInt(line[c]);
            }
        }
        visit = new boolean[R][C];
        idx = 2;

        // 네이밍
        for(int r = 0; r<R; r++){
            for(int c = 0; c<C; c++){
                if( map[r][c] == 1){
                    naming(r,c,idx++);
                }
            }
        }

        // 거리 구하기

        // adj 배열 초기화
        adj = new int[idx][idx];
        for(int r = 0; r < idx; r++){
            adj[r] = new int[idx];
            for(int c = 0; c < idx; c++){
                adj[r][c] = 10;
            }
        }
        visit = new boolean[R][C];
        for(int r = 0; r < R ; r++){
            for(int c = 0; c < C; c++){
                if(map[r][c] != 0 && ! visit[r][c]){
                    check_dist(r,c,map[r][c]);
                }
            }
        }

        // adj 배열을 ( start, end, distance ) 로 변환
        dists = new LinkedList<>();
        for(int r = 2; r < idx ; r++){
            for(int c = r+1; c < idx; c++){
                if(adj[r][c] == 10) continue;
                dists.add(new Dist(r,c,adj[r][c]));
            }
        }

        dists.sort(Dist::compareTo);

        // 최소 스패닝 트리
        int[] pos = new int[idx];
        for(int i = 0; i<idx; i++)
            pos[i] = -1;

        LinkedList<LinkedList<Integer>> boxes = new LinkedList<>();
        LinkedList<Integer> box;

        int answer = 0;
        for(Dist dist : dists){

            int a = dist.s;
            int b = dist.e;
            int d = dist.d;

            if(pos[a] == -1 && pos[b] == -1){
                // a, b 가 어디에도 속하지 않음
                box = new LinkedList<>();
                box.add(a);
                box.add(b);

                pos[a] = boxes.size();
                pos[b] = boxes.size();

                boxes.add(box);
            }else if(pos[a] == -1 || pos[b] == -1) {
                // 둘 중 하나가 속하지 않음
                int yesG = (pos[a] == -1) ? b : a;
                int noG = (pos[a] == -1) ? a : b;
                boxes.get(pos[yesG]).add(noG);
                pos[noG] = pos[yesG];
            }else{
                // a, b 모두 속함.
                // 같을때
                if(pos[a] == pos[b]){
                    continue;
                }
                // 다를때 병합.
                int sIdx = (pos[a] > pos[b]) ? b : a;
                int bIdx = (pos[a] > pos[b]) ? a : b;
                int sPos = pos[sIdx];
                int bPos = pos[bIdx];

                // bidx 에서 sidx 로 옯김
                for(Integer bo : boxes.get(bPos)){
                    boxes.get(sPos).add(bo);    // 이사 가고
                    pos[bo] = sPos;             // 구청에 신고 하고
                }
                boxes.get(bPos).clear();        // 이전 집 정리
            }

            answer += d;
        }
        if(boxes.size() == 0 || boxes.get(0).size() != idx-2)
            System.out.println("-1");
        else
            System.out.println(answer);
    }
    static void print_map(){
        for(int r = 0; r<R; r++){
            for(int c = 0; c<C; c++){
                System.out.print(map[r][c] + " " );
            }
            System.out.println();
        }
        System.out.println();
    }

    static void print_adj(){
        for(int r = 2; r<idx; r++){
            for(int c = 2; c<idx; c++){
                System.out.print(adj[r][c] + " " );
            }
            System.out.println();
        }
        System.out.println();
    }
    static void naming(int r, int c, int idx){
        visit[r][c] = true;
        map[r][c] = idx;
        for(int[] d : dir){
            int rr = r + d[0];
            int cc = c + d[1];
            if(0 <= rr && rr < R && 0 <= cc && cc < C && map[rr][cc] == 1){
                naming(rr,cc,idx);
            }
        }
    }

    static void check_dist(int r,int c, int name){
        visit[r][c] = true;
        Queue<Loc> que ; // 현재 레벨 큐
        Queue<Loc> que_; // 다음 레벨 큐

        LinkedList<Loc> group = new LinkedList<>(); // name 집합

        que = new LinkedList<>();
        que.add(new Loc(r,c));

        while(! que.isEmpty()){
            que_ = new LinkedList<>();
            for(Loc l : que){
                group.add(l);
                for(int[] d : dir){
                    int rr = l.r + d[0];
                    int cc = l.c + d[1];

                    if(0 <= rr && rr < R && 0 <= cc && cc < C && map[rr][cc] == name && ! visit[rr][cc]){
                        visit[rr][cc] = true;
                        que_.add(new Loc(rr,cc));
                    }
                }
            }
            que = que_;
        }


        que = new LinkedList<>();
        for(Loc l : group){
            for(int d = 0; d < 4; d++){
                que.add(new Loc(l.r, l.c, d));
            }
        }

        // ======현재 섬부터 다른 섬까지의 거리 구하기=========================
        int distance = 0;
        while (! que.isEmpty()){
            que_ = new LinkedList<>();

            for(Loc l : que){
                int rr = l.r + dir[l.d][0];
                int cc = l.c + dir[l.d][1];

                if(0 <= rr && rr < R && 0 <= cc && cc < C ) {
                    if (map[rr][cc] == name)
                        continue;
                    else if (map[rr][cc] == 0) {
                        l.r = rr;
                        l.c = cc;
                        que_.add(l);
                    } else {  // 다른 섬 발견
                        if(distance < 2) continue;
                        if (adj[name][map[rr][cc]] == 10)
                            adj[name][map[rr][cc]] = distance;
                    }
                }
            }
            que = que_;
            distance ++;
        }

    }
}
